op-e2e/interop/supersystem.go

```diff
@@ -0,0 +1,616 @@
+	"github.com/ethereum-optimism/optimism/op-service/oppprof"
+	oprpc "github.com/ethereum-optimism/optimism/op-service/rpc"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	supervisorConfig "github.com/ethereum-optimism/optimism/op-supervisor/config"
+	"github.com/ethereum-optimism/optimism/op-supervisor/supervisor"
+)
+
+// SuperSystem is an interface for the system (collection of connected resources)
+// it provides a way to get the resources for a network by network ID
+// and provides a way to get the list of network IDs
+// this is useful for testing multiple network backends,
+// for example, interopE2ESystem is the default implementation, but a shim to
+// kurtosis or another testing framework could be implemented
+type SuperSystem interface {
+	// get the supervisor
+	Supervisor() *supervisor.SupervisorService
+	// get the supervisor client
+	SupervisorClient() *sources.SupervisorClient
+	// get the batcher for a network
+	Batcher(id *big.Int) *bss.BatcherService
+	// get the proposer for a network
+	Proposer(id *big.Int) *l2os.ProposerService
+	// get the opnode for a network
+	OpNode(id *big.Int) *opnode.Opnode
+	// get the geth instance for a network
+	L2Geth(id *big.Int) *geth.GethInstance
+	// get the L2 geth client for a network
+	L2GethClient(id *big.Int) *ethclient.Client
+	// get the secret for a network and role
+	L2OperatorKey(id *big.Int, role devkeys.ChainOperatorRole) ecdsa.PrivateKey
+	// get the list of network IDs
+	L2IDs() []*big.Int
+	// register a username to an account on all L2s
+	AddUser(username string)
+	// get the user key for a user on an L2
+	UserKey(id *big.Int, username string) ecdsa.PrivateKey
+	// send a transaction on an L2 on the given network, from the given user
+	SendL2Tx(id *big.Int, username string, applyTxOpts op_e2e.TxOptsFn) *types.Receipt
+	// get the address for a user on an L2
+	Address(id *big.Int, username string) common.Address
+}
+
+// NewSuperSystem creates a new SuperSystem from a recipe. It creates an interopE2ESystem.
+func NewSuperSystem(t *testing.T, recipe *interopgen.InteropDevRecipe, w worldResourcePaths) SuperSystem {
+	s2 := &interopE2ESystem{recipe: recipe}
+	s2.prepare(t, w)
+	return s2
+}
+
+// interopE2ESystem implements the SuperSystem interface
+// it prepares network resources and provides access to them
+// the functionality is broken down into smaller functions so that
+// the system can be prepared iteratively if desired
+type interopE2ESystem struct {
+	t               *testing.T
+	recipe          *interopgen.InteropDevRecipe
+	logger          log.Logger
+	hdWallet        *devkeys.MnemonicDevKeys
+	worldDeployment *interopgen.WorldDeployment
+	worldOutput     *interopgen.WorldOutput
+	beacon          *fakebeacon.FakeBeacon
+	l1              *geth.GethInstance
+	l2s             map[*big.Int]l2Set
+	l2GethClients   map[*big.Int]*ethclient.Client
```

Comment:
*big.Int is a pointer, not safe to use as map key. And the non-pointer version of `*big.Int` is a slice, also not safe to use.

Hence I thought the previous `string` was better, cc @tynes.

Another option, as I do in some of the interop deployment code, is to use `uint256.Int` (the raw value, not a pointer to it), which is a typed 4-uint64 array, save to use for raw Go value comparisons.

Reply:
I did consider that a slice/pointer would not be safe to use, but that's only if the value is changing over time, right? Since they're static keys, they shouldn't get reassigned.

Reply:
Talked with @protolambda offline, we're not happy with the unsafety of a big int pointer as a key, even if the value does not change. I'll revert back to a `string`